src/tracks/playabletrack/wavetrack/ui/SampleHandle.cpp

```diff
@@ -345,66 +361,25 @@ UIHandle::Result SampleHandle::Drag
    if (mAltKey)
       return RefreshNone;
 
-   sampleCount s0;     //declare this for use below.  It designates which sample number to draw.
-
-   // Figure out what time the click was at
-   //Find the point that we want to redraw at. If the control button is down,
-   //adjust only the originally clicked-on sample
-
-   if (event.m_controlDown) {
-      //*************************************************
-      //***   CTRL-DOWN (Hold Initial Sample Constant ***
-      //*************************************************
-
-      s0 = mClickedStartSample;
-   }
-   else {
-      //*************************************************
-      //***    Normal CLICK-drag  (Normal drawing)    ***
-      //*************************************************
-
-      //Otherwise, adjust the sample you are dragging over right now.
-      //convert this to samples
-      const double tt = viewInfo.PositionToTime(event.m_x, mRect.x);
-      s0 = mClickedTrack->TimeToLongSamples(tt);
-   }
-
-   const double t0 = mClickedTrack->LongSamplesToTime(s0);
-
-   // Do redrawing, based on the mouse position.
-   // Calculate where the mouse is located vertically (between +/- 1)
-
+   const double t0 = viewInfo.PositionToTime(mLastDragPixel);
+   const double t1 = viewInfo.PositionToTime(event.m_x, mRect.x);
+   const int x1 =
+      event.m_controlDown ? mClickedStartPixel : viewInfo.TimeToPosition(t1);
    const float newLevel = FindSampleEditingLevel(event, viewInfo, t0);
+   const auto start = std::min(t0, t1);
+   const auto end = std::max(t0, t1);
+   // For fast pencil movements covering more than one sample between two
+   // updates, we draw a line going from v0 at t0 to v1 at t1.
+   const auto interpolator = [t0, t1, v0 = mLastDragSampleValue,
+                              v1 = newLevel](double t) {
+      const auto gradient = (v1 - v0) / (t1 - t0);
```

Comment:
Does it ever divide by 0?

Reply:
Good question. The code before didn't check, but I cannot guarantee that my changes do not change that particular thing.
Adding a check upstream.